### PR TITLE
[SPARK-33130][SQL] Support ALTER TABLE in JDBC v2 Table Catalog: add, update type and nullability of columns (MsSqlServer dialect)

### PR DESCRIPTION
**Pull request by ScrapCodes**
### What changes were proposed in this pull request?

Override the default SQL strings for:
ALTER TABLE RENAME COLUMN
ALTER TABLE UPDATE COLUMN NULLABILITY
in the following MsSQLServer JDBC dialect according to official documentation.
Write MsSqlServer integration tests for JDBC.

### Why are the changes needed?

To add the support for alter table when interacting with MSSql Server.

### How was this patch tested?

added tests

### CODE REVIEW

**Comment by huaxingao** [comment]
Please rebase to get the latest change for ```JdbcDialects.scala```.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+package org.apache.spark.sql.jdbc.v2
+
+import java.sql.Connection
+import java.sql.SQLFeatureNotSupportedException
+
+import org.scalatest.time.SpanSugar._
+
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.parser.ParseException
```
Comment (huaxingao) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:27]: not needed?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import org.apache.spark.sql.execution.datasources.v2.jdbc.JDBCTableCatalog
+import org.apache.spark.sql.jdbc.{DatabaseOnDocker, DockerJDBCIntegrationSuite}
+import org.apache.spark.sql.test.SharedSparkSession
```
Comment (huaxingao) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:30]: not needed?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+  override def sparkConf: SparkConf = super.sparkConf
+    .set("spark.sql.catalog.mssql", classOf[JDBCTableCatalog].getName)
+    .set("spark.sql.catalog.mssql.url", db.getJdbcUrl(dockerIp, externalPort))
+
+  override val connectionTimeout = timeout(7.minutes)
+
+  override def dataPreparation(conn: Connection): Unit = {}
+
+  test("ALTER TABLE - rename column") {
```
Comment (huaxingao) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:84]: Could you please move this test to the parent class? I think other databases need to test this one too.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala [hunk 4]
```diff
@@ -64,4 +65,39 @@ private object MsSqlServerDialect extends JdbcDialect {
...
   }
+
+  // scalastyle:off line.size.limit
+  // see https://docs.microsoft.com/en-us/sql/relational-databases/tables/add-columns-to-a-table-database-engine?view=sql-server-ver15
+  // scalastyle:on line.size.limit
+  override def getAddColumnQuery(
+      tableName: String,
+      columnName: String,
+      dataType: String): String = {
+    s"ALTER TABLE $tableName ADD $columnName $dataType"
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:76]: $columnName -> ${quoteIdentifier(columnName)}
please see #30041

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala [hunk 5]
```diff
@@ -64,4 +65,39 @@ private object MsSqlServerDialect extends JdbcDialect {
...
+  }
+
+  // scalastyle:off line.size.limit
+  // See https://docs.microsoft.com/en-us/sql/relational-databases/system-stored-procedures/sp-rename-transact-sql?view=sql-server-ver15
+  // scalastyle:on line.size.limit
+  override def getRenameColumnQuery(
+      tableName: String,
+      columnName: String,
+      newName: String): String = {
+    s"EXEC sp_rename '$tableName.$columnName', '$newName', 'COLUMN'"
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/MsSqlServerDialect.scala:86]: '$tableName.$columnName' -> '$tableName.${quoteIdentifier(columnName)}
'$newName'->${quoteIdentifier(newName)}

**Comment by ScrapCodes** [comment]
This needs more work, while testing this, I found rename column needs some work for mysql dialect. I am currently working on a patch for the same.

**Comment by ScrapCodes** [comment]
Hi @huaxingao and @cloud-fan , This is ready for review !

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.jdbc.v2
+
+import java.sql.Connection
+import java.sql.SQLFeatureNotSupportedException
```
Comment (huaxingao) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:21]: nit: combine these two lines?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      "ACCEPT_EULA" -> "Y"
+    )
+    override val usesIpc = false
+    override val jdbcPort: Int = 1433
+
+    override def getJdbcUrl(ip: String, port: Int): String =
+      s"jdbc:sqlserver://$ip:$port;user=sa;password=Sapass123;"
+  }
+
+  override def testUpdateColumnType(tbl: String): Unit = {
```
Comment (huaxingao) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:58]: nit: shall we put this method after `def dataPreparation`?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  override def sparkConf: SparkConf = super.sparkConf
+    .set("spark.sql.catalog.mssql", classOf[JDBCTableCatalog].getName)
+    .set("spark.sql.catalog.mssql.url", db.getJdbcUrl(dockerIp, externalPort))
+
+  override val connectionTimeout = timeout(7.minutes)
+
+  override def dataPreparation(conn: Connection): Unit = {}
+
+  override def testUpdateColumnNullability(tbl: String): Unit = {
+    sql("CREATE TABLE mssql.alt_table (ID STRING NOT NULL) USING _")
```
Comment (huaxingao) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:83]: ```mssql.alt_table``` ->  ```$tbl```?

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    // Update nullability is unsupported for mssql db.
+    val msg = intercept[AnalysisException] {
+      sql("ALTER TABLE mssql.alt_table ALTER COLUMN ID DROP NOT NULL")
```
Comment (huaxingao) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/v2/MsSqlServerIntegrationSuite.scala:86]: ditto

**Comment by huaxingao** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ScrapCodes** [comment]
Thanks !!

**Pull request closed without merging**